File: src/data/minigamePacks.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import {
  MINI_GAME_PACKS,
  scoreOrderedSteps,
  scoreRedFlagSort,
  shuffledStepIds,
  validateMiniGamePack,
  validMiniGamePacks,
  type OrderedMiniGamePack,
  type RedFlagSortPack,
} from './minigamePacks.ts'

test('every local mini-game pack is explicitly reviewed and valid', () => {
  assert.ok(MINI_GAME_PACKS.length >= 3)
  for (const pack of MINI_GAME_PACKS) {
    assert.deepEqual(validateMiniGamePack(pack), [])
    assert.ok(pack.source.label)
    assert.ok(pack.source.reviewedBy)
  }
})

test('each requested authored medicine game has a valid local pack', () => {
  assert.ok(validMiniGamePacks('clinical_sequence').length >= 1)
  assert.ok(validMiniGamePacks('mechanism_chain').length >= 1)
  assert.ok(validMiniGamePacks('red_flag_sort').length >= 1)
})

test('ordered games shuffle deterministically without changing authored answers', () => {
  const pack = validMiniGamePacks('clinical_sequence')[0] as OrderedMiniGamePack
  const a = shuffledStepIds(pack, 12)
  const b = shuffledStepIds(pack, 12)
  assert.deepEqual(a, b)
  assert.deepEqual([...a].sort(), pack.steps.map((step) => step.id).sort())
  assert.deepEqual(scoreOrderedSteps(pack, pack.steps.map((step) => step.id)), {
    exactPositions: pack.steps.length,
    total: pack.steps.length,
    complete: true,
    correct: pack.steps.map((step) => step.id),
  })
})

test('ordered scoring reports exact-position matches only', () => {
  const pack = validMiniGamePacks('mechanism_chain')[0] as OrderedMiniGamePack
  const submitted = pack.steps.map((step) => step.id)
  submitted.reverse()
  const score = scoreOrderedSteps(pack, submitted)
  assert.equal(score.total, pack.steps.length)
  assert.ok(score.exactPositions < pack.steps.length)
  assert.equal(score.complete, true)
})

test('red flag sort scoring uses authored lanes and rationales', () => {
  const pack = validMiniGamePacks('red_flag_sort')[0] as RedFlagSortPack
  const placements = Object.fromEntries(pack.findings.map((finding) => [finding.id, finding.lane]))
  const score = scoreRedFlagSort(pack, placements)
  assert.equal(score.correct, pack.findings.length)
  assert.equal(score.total, pack.findings.length)
  assert.ok(score.results.every((result) => result.correct && result.finding.rationale.length > 0))
})

test('invalid packs are refused by validation before the UI can use them', () => {
  const pack: OrderedMiniGamePack = {
    id: 'bad',
    kind: 'clinical_sequence',
    title: '',
    subjectId: '',
    topic: 'Nowhere',
    summary: 'Broken fixture',
    prompt: 'Broken',
    source: { label: '', reviewedBy: '' },
    steps: [{ id: 'x', text: '' }],
    explanation: '',
  }
  const errors = validateMiniGamePack(pack)
  assert.ok(errors.length >= 4)
})

File: android/app/src/test/java/com/synapse/app/core/minigames/MinigamePacksTest.kt
package com.synapse.app.core.minigames

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/** Ported from web's `src/data/minigamePacks.test.ts`, same assertions, same fixtures. */
class MinigamePacksTest {

    @Test
    fun everyLocalPackIsExplicitlyReviewedAndValid() {
        assertTrue(MINI_GAME_PACKS.size >= 3)
        for (pack in MINI_GAME_PACKS) {
            assertEquals(emptyList<String>(), validateMiniGamePack(pack))
            assertTrue(pack.source.label.isNotBlank())
            assertTrue(pack.source.reviewedBy.isNotBlank())
        }
    }

    @Test
    fun eachRequestedAuthoredMedicineGameHasAValidLocalPack() {
        assertTrue(validMiniGamePacks(MiniGameKind.CLINICAL_SEQUENCE).isNotEmpty())
        assertTrue(validMiniGamePacks(MiniGameKind.MECHANISM_CHAIN).isNotEmpty())
        assertTrue(validMiniGamePacks(MiniGameKind.RED_FLAG_SORT).isNotEmpty())
    }

    @Test
    fun orderedGamesShuffleDeterministicallyWithoutChangingAuthoredAnswers() {
        val pack = validMiniGamePacks(MiniGameKind.CLINICAL_SEQUENCE).first() as OrderedMiniGamePack
        val a = shuffledStepIds(pack, 12)
        val b = shuffledStepIds(pack, 12)
        assertEquals(a, b)
        assertEquals(pack.steps.map { it.id }.sorted(), a.sorted())
        assertEquals(
            OrderedStepScore(
                exactPositions = pack.steps.size,
                total = pack.steps.size,
                complete = true,
                correct = pack.steps.map { it.id },
            ),
            scoreOrderedSteps(pack, pack.steps.map { it.id }),
        )
    }

    /**
     * Exact shuffle output for seed 12 over the clinical-sequence pack,
     * cross-checked against web's `seededRandom`/`shuffle` run on the same
     * pack and seed via Node — pins the Kotlin port to the identical
     * bit-for-bit sequence, not merely "a" deterministic one.
     */
    @Test
    fun shuffledStepIdsMatchesTheWebReferenceSequenceForSeedTwelve() {
        val pack = validMiniGamePacks(MiniGameKind.CLINICAL_SEQUENCE).first() as OrderedMiniGamePack
        assertEquals(
            listOf("danger", "compressions", "response", "breathing", "airway", "aed"),
            shuffledStepIds(pack, 12),
        )
    }

    @Test
    fun orderedScoringReportsExactPositionMatchesOnly() {
        val pack = validMiniGamePacks(MiniGameKind.MECHANISM_CHAIN).first() as OrderedMiniGamePack
        val submitted = pack.steps.map { it.id }.reversed()
        val score = scoreOrderedSteps(pack, submitted)
        assertEquals(pack.steps.size, score.total)
        assertTrue(score.exactPositions < pack.steps.size)
        assertTrue(score.complete)
    }

    @Test
    fun redFlagSortScoringUsesAuthoredLanesAndRationales() {
        val pack = validMiniGamePacks(MiniGameKind.RED_FLAG_SORT).first() as RedFlagSortPack
        val placements = pack.findings.associate { it.id to it.lane }
        val score = scoreRedFlagSort(pack, placements)
        assertEquals(pack.findings.size, score.correct)
        assertEquals(pack.findings.size, score.total)
        assertTrue(score.results.all { it.correct && it.finding.rationale.isNotEmpty() })
    }

    @Test
    fun invalidPacksAreRefusedByValidationBeforeTheUiCanUseThem() {
        val pack = OrderedMiniGamePack(
            id = "bad",
            kind = MiniGameKind.CLINICAL_SEQUENCE,
            title = "",
            subjectId = "",
            topic = "Nowhere",
            summary = "Broken fixture",
            prompt = "Broken",
            source = MiniGameSource(label = "", reviewedBy = ""),
            steps = listOf(OrderedStep("x", "")),
            explanation = "",
        )
        val errors = validateMiniGamePack(pack)
        assertTrue(errors.size >= 4)
    }
}
